pkg/runtime/local/local.go

```diff
@@ -0,0 +1,169 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package local
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+)
+
+type Runtime struct{}
+
+func (r *Runtime) Init(runtimeParams *params.Params) error {
+	return nil
+}
+
+func (r *Runtime) Close() error {
+	return nil
+}
+
+func (r *Runtime) GlobalParamDescs() params.ParamDescs {
+	return nil
+}
+
+func (r *Runtime) RunGadget(
+	gadgetCtx runtime.GadgetContext,
+) ([]byte, error) {
+	logger := gadgetCtx.Logger()
+
+	logger.Debugf("running with local runtime")
+
+	gadgetInst, ok := gadgetCtx.GadgetDesc().(gadgets.GadgetInstantiate)
+	if !ok {
+		return []byte{}, errors.New("gadget not instantiable")
+	}
+
+	return r.runGadget(gadgetCtx, gadgetInst)
+}
+
+func (r *Runtime) runGadget(
+	gadgetCtx runtime.GadgetContext,
+	gadget gadgets.GadgetInstantiate,
+) (out []byte, err error) {
+	log := gadgetCtx.Logger()
+
+	operatorsParamCollection := gadgetCtx.OperatorsParamCollection()
+
+	// Create gadget instance
+	gadgetInstance, err := gadget.NewInstance(gadgetCtx)
+	if err != nil {
+		return out, fmt.Errorf("instantiating gadget: %w", err)
+	}
+
+	// Deferring getting results and closing to make sure operators got their chance to clean up properly beforehand
+	defer func() {
+		if closer, ok := gadgetInstance.(gadgets.CloseGadget); ok {
+			log.Debugf("calling gadget.Close()")
+			closer.Close()
+		}
+
+		// No need to get results if gadget failed
+		if err != nil {
+			return
+		}
+
+		if results, ok := gadgetInstance.(gadgets.GadgetResult); ok {
+			log.Debugf("getting result")
+			out, err = results.Result()
+			if err != nil {
+				err = fmt.Errorf("getting result: %w", err)
+			}
+		}
+	}()
+
+	// Install operators
+	operatorInstances, err := gadgetCtx.Operators().Instantiate(gadgetCtx, gadgetInstance, operatorsParamCollection)
+	if err != nil {
+		return out, fmt.Errorf("instantiating operators: %w", err)
+	}
+	defer operatorInstances.Close()
+	log.Debugf("found %d operators", len(gadgetCtx.Operators()))
+
+	if gadget.Type() == gadgets.TypeTraceIntervals {
+		// Enable interval pushes
+		interval := gadgetCtx.GadgetParams().Get(gadgets.ParamInterval).AsInt()
+		log.Debugf("enabling snapshots every %d seconds", interval)
+		gadgetCtx.Parser().EnableSnapshots(gadgetCtx.Context(), time.Second, interval)
+	}
```

Comment:
I don't think we need the snapshots for the local runtime. In this case the gadget will produce the result each interval seconds and will call the callback. All the information comes from the same place (a single gadget), hence there is synchronization issue as in kubectl-gadget, i.e, the snapshot combiner is not needed for this one.

Reply:
This still has to be addressed.

Reply:
Removed.

Reply:
Do you think we can get rid of EnableSnapshots and snapshot combiner too? (trying to keep this PR as small as possible).

Reply:
Hmm, sure, we could remove it, but it would just move it to the next PR and we would have to review it again, there (especially the snapshot combiner, not so much concerned about the EnableSnapshots).

Reply:
TBH I haven't reviewed that part in detail because I focused in other things. I'm not sure if somebody actually checked it, I see a comment from @eiffel-fl and that's it.